### PR TITLE
[D2D-50] Adapts user to work on both Jira Cloud and Jira Server

### PR DESCRIPTION
**Pull request by taksan**
- Refactoring JiraUser to abstract the user identification and fixes all places that depends on identification, such as Assignees, Avatar, Co  Assignees, Comments, and so on
- Fixes roles matching by using the new user identification

### CODE REVIEW

File: test-support/src/main/java/objective/taskboard/testUtils/JiraMockServer.java [hunk 1]
```diff
@@ -552,18 +562,22 @@ private static JSONObject createEmptyStatus() throws JSONException {
...
         }
         coassigneeField.put(aKey, coassignees);
     }
 
-    private static void setAssignee(JSONObject fields, JSONObject each) throws JSONException {
-        JSONObject makeAssignee = createEmptyAssignee();
-        makeAssignee.put("name", each.get("name"));
-        fields.put("assignee", makeAssignee);
+    private void setAssignee(JSONObject fields, JSONObject each) throws JSONException {
+        if ((""+each.getString("name")).equals("null")) {
```
Comment (herbert-venancio) [inline test-support/src/main/java/objective/taskboard/testUtils/JiraMockServer.java]: getString throws exception if name doesn't exist, maybe change to:
```Java
if ("null".equals(""+each.opt("name"))) {
```

File: core/src/test/java/objective/taskboard/jira/client/JiraIssueDtoSearchTest.java [hunk 2]
```diff
@@ -11,9 +11,19 @@
...
 
+import objective.taskboard.jira.data.JiraUser;
+import objective.taskboard.utils.SpringContextBridgeTestSupport;
+
 public class JiraIssueDtoSearchTest {
+    
+    @Before
+    public void setup() {
+        SpringContextBridgeTestSupport.setupForServerWithEmptyProperties();
+    }
```
Comment (herbert-venancio) [inline core/src/test/java/objective/taskboard/jira/client/JiraIssueDtoSearchTest.java]: Add the teardown counterpart:
```Java
    @After
    public void disposeSpringContextBridge() {
        SpringContextBridgeTestSupport.teardown();
    }
```

File: core/src/test/java/objective/taskboard/utils/SpringContextBridgeTestSupport.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    public static void setupForServerWithEmptyProperties() {
+        setupForServerGivenProperties(defaultProperties());
+    }
+
+    public static void setupForServerGivenProperties(JiraClientProperties properties) {
+        ApplicationContext ac = mock(ApplicationContext.class);
+        when(ac.getBean(JiraClientProperties.class)).thenReturn(properties);
+        new SpringContextBridge().setApplicationContext(ac);
+    }
+
```
Comment (herbert-venancio) [inline core/src/test/java/objective/taskboard/utils/SpringContextBridgeTestSupport.java]: Add a `teardown` method to help with tests:
```Java
    public static void teardown() {
        new SpringContextBridge().setApplicationContext(null);
    }
```

File: core/src/main/java/objective/taskboard/auth/TaskboardProjectRolesResolver.java [hunk 4]
```diff
@@ -32,16 +32,16 @@ public TaskboardProjectRolesResolver(ProjectService projectService, JiraService
...
+        return getUserJiraRoles().stream()
                 .flatMap(r -> 
                     projectService.getTaskboardProjectsForJiraProjectKey(r.projectKey).stream()
                         .map(pk -> new ProjectRole(r.name, pk.getProjectKey())))
                 .collect(toList());
     }
     
-    List<Role> getUserJiraRoles(String username) {
-        return userCache.computeIfAbsent(USER_JIRA_ROLES, () -> jiraService.getUserRoles(username));
+    List<Role> getUserJiraRoles() {
```
Comment (herbert-venancio) [inline core/src/main/java/objective/taskboard/auth/TaskboardProjectRolesResolver.java]: rename to get**Logged**UserJiraRoles

File: core/src/test/java/objective/taskboard/jira/JiraServiceTest.java [hunk 5]
```diff
@@ -147,149 +152,14 @@ public void givenJiraServer_whenGetExistentJiraUser_shouldReturnJiraUser() {
 
     @Test
     public void givenJiraServer_whenGetNonexistentJiraUser_shouldReturnNull() {
```
Comment (herbert-venancio) [inline core/src/test/java/objective/taskboard/jira/JiraServiceTest.java]: rename to givenJiraServer_whenGetNonexistentJiraUser_should**ThrowUserNotFoundException**

File: core/src/test/java/objective/taskboard/testUtils/CredentialHolderUtils.java [hunk 6]
```diff
@@ -27,6 +29,10 @@ public static void mockLoggedUser(String username, String password) {
         doReturn(username).when(loggedInUser).getRealUsername();
         doReturn(Optional.empty()).when(loggedInUser).getImpersonateUsername();
         doReturn(loggedInUser).when(authentication).getPrincipal();
+        User user = new User(username, "display name", "email");
+       
+        //doReturn(user).when(loggedInUser).getUser();
```
[the remark this answers is not in the document]
Comment (taksan) [reply core/src/test/java/objective/taskboard/testUtils/CredentialHolderUtils.java]: done

File: jira-client/pom.xml [hunk 7]
```diff
@@ -82,5 +82,10 @@
 			<artifactId>sonar-jacoco-listeners</artifactId>
 			<scope>test</scope>
 		</dependency>
+		<dependency>
+            <groupId>org.mockito</groupId>
```
[the remark this answers is not in the document]
Comment (taksan) [reply jira-client/pom.xml]: done

File: jira-client/src/main/java/objective/taskboard/jira/data/JiraUser.java [hunk 8]
```diff
@@ -8,40 +8,47 @@
...
+    private static String S24_24 = "24x24";
+    private static String S48_48 = "48x48";
 
-    public String name;
-    public String displayName;
-    public String emailAddress;
-    public Map<String, URI> avatarUrls = emptyMap();
-    public String locale;
-    public GroupList groups;
+    protected String self;
```
[the remark this answers is not in the document]
Comment (taksan) [reply jira-client/src/main/java/objective/taskboard/jira/data/JiraUser.java]: done

File: jira-client/src/main/java/objective/taskboard/jira/data/JiraUser.java [hunk 9]
```diff
@@ -84,22 +91,42 @@ public GroupList deserialize(JsonParser p, DeserializationContext ctxt) throws I
...
 
-        @GET("/rest/api/latest/myself")
+        @GET("/rest/api/latest/myself?expand=groups")
         JiraUser myself();
 
         @GET("/rest/api/latest/user/search")
         List<JiraUser> findUsers(@Query("username") String username);
+        
+        @GET("/rest/api/latest/user/search")
+        List<JiraUser> findUsersByAccountId(@Query("accountId") String username);
```
[the remark this answers is not in the document]
Comment (taksan) [reply jira-client/src/main/java/objective/taskboard/jira/data/JiraUser.java]: done

File: pom.xml [hunk 10]
```diff
@@ -297,6 +297,12 @@
 				<artifactId>spring-security-web</artifactId>
 				<version>${spring-security.version}</version>
 			</dependency>
+	        <dependency>
```
[the remark this answers is not in the document]
Comment (taksan) [reply pom.xml]: fixed

File: application/src/main/resources/application-dev.properties [hunk 11]
```diff
@@ -49,33 +49,33 @@ jira.issuetype.subtasks[0].id=
...
-jira.resolutions.done.name=Done
-jira.resolutions.canceled.name=Canceled
+jira.resolutions.done.name=
+jira.resolutions.canceled.name=
 
 # main tshirt custom field
-jira.customfield.t-shirt-size.mainTShirtSizeFieldId=customfield_11457
+jira.customfield.t-shirt-size.mainTShirtSizeFieldId=
 
 # Resolvido, Fechado, Cancelado, Or�ado, Done, Recebido
```
[the remark this answers is not in the document]
Comment (taksan) [reply application/src/main/resources/application-dev.properties]: Done

File: core/src/main/java/objective/taskboard/issueBuffer/IssueFacade.java [hunk 12]
```diff
@@ -135,7 +135,7 @@ public Issue removeAssigneeFromIssue(String issueKey, String username) {
 
         List<User> assignees = new ArrayList<>(issue.getAssignees());
 
-        assignees.removeIf(anAssignee -> anAssignee.isAssigned() && anAssignee.name.equals(username));
+        assignees.removeIf(anAssignee -> anAssignee.isAssigned() && anAssignee.userId.equals(username));
```
[the remark this answers is not in the document]
Comment (taksan) [reply core/src/main/java/objective/taskboard/issueBuffer/IssueFacade.java]: done

File: core/src/test/java/objective/taskboard/testUtils/CredentialHolderUtils.java [hunk 13]
```diff
@@ -10,10 +11,11 @@
 import org.springframework.security.core.context.SecurityContextHolder;
 
 import objective.taskboard.auth.LoggedUserDetails;
+import objective.taskboard.data.User;
 
 public class CredentialHolderUtils {
 
-    public static void mockLoggedUser(String username, String password) {
+    public static void mockLoggedUser(String username, String password, String deploymentType) {
```
[the remark this answers is not in the document]
Comment (taksan) [reply core/src/test/java/objective/taskboard/testUtils/CredentialHolderUtils.java]: done

File: sizing-importer/src/main/java/objective/taskboard/google/SpreadsheetsManager.java [hunk 14]
```diff
@@ -146,6 +146,11 @@ public SpreadsheeNotFoundException(String spreadsheetId) {
 
     @ResponseStatus(value = HttpStatus.SERVICE_UNAVAILABLE, reason="We are having communication issues with Google Spreadsheet. Please, try again in 15 minutes.")
     public static class SpreadsheeBackendException extends SpreadsheetException {
```
[the remark this answers is not in the document]
Comment (taksan) [reply sizing-importer/src/main/java/objective/taskboard/google/SpreadsheetsManager.java]: done

**Pull request merged**